src/Session.cc

```diff
@@ -301,10 +301,14 @@ void Session::OnRegisterViewer(const CARTA::RegisterViewer& message, uint16_t ic
 }
 
 void Session::OnFileListRequest(const CARTA::FileListRequest& request, uint32_t request_id) {
+    auto progress_callback = [&](CARTA::Progress progress) { SendEvent(CARTA::EventType::FILE_LIST_PROGRESS, request_id, progress); };
+    _file_list_handler->SetProgressCallBack(progress_callback);
     CARTA::FileListResponse response;
     FileListHandler::ResultMsg result_msg;
     _file_list_handler->OnFileListRequest(request, response, result_msg);
-    SendEvent(CARTA::EventType::FILE_LIST_RESPONSE, request_id, response);
+    if (!response.cancel()) {
```

Comment:
@kswang1029 I wonder if we should still return the _partial_ results, even after cancellation? At least then users can see some of the files or directories. In fact, if we do this, we can also move the directories first in order to prioritise them

Reply:
we could do that as long as we warn user well that what they see is not complete. The warning text should be persistent somewhere on UI I think.

Reply:
right, it could be shown in the folder name. e.g scratch -> images -> lots _(partial results)_.

I think it would be particularly helpful for navigating to a subdirectory

Reply:
If we loaded an image from the partial list then try to append a new, when we launch the file browser will it request the full list again?

Reply:
good question...

Reply:
I remembered the frontend does request the file list again (and it should instead of using cached list). So I think we are fine.

Reply:
So should the backend return the partial file list after the cancelation? If the frontend caches a file list, next time it can tell the backend to get a life list started from the *n-th* file or folder.

Reply:
No, next time it should still start from scratch because the file list may change